core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala

```diff
@@ -0,0 +1,719 @@
+    nowTimestamp.flatMap { now =>
+      inTransactionT { _ =>
+        for {
+          job <- storageTransferJobs
+            .filter(_.finished.isEmpty)
+            .sortBy(_.updated.asc)
+            .take(1)
+            .result
+            .headOption
+
+          // touch the job so that next call to peek returns another
+          _ <- if (job.isDefined)
+            storageTransferJobs.filter(_.id === job.get.id).map(_.updated).update(now) else
+            DBIO.successful()
+
+        } yield job.map(_.copy(updated = now))
+      }
+    }
+
+
+  final def getGoogleProjectAndTmpBucket(migration: WorkspaceMigration)
+  : MigrateAction[(GoogleProjectId, GcsBucketName)] =
+    MigrateAction.liftIO(IO {
+      (
+        migration.newGoogleProjectId.getOrElse(throw noGoogleProjectError(migration)),
+        migration.tmpBucketName.getOrElse(throw noTmpBucketError(migration))
+      )
+    })
+
+
+  final def inTransactionT[A](action: DataAccess => ReadWriteAction[Option[A]]): MigrateAction[A] =
+    inTransaction(action).mapF(optT => OptionT(optT.value.map(_.flatten)))
+
+
+  final def inTransaction[A](action: DataAccess => ReadWriteAction[A]): MigrateAction[A] =
+    for {
+      dataSource <- MigrateAction.asks(_.dataSource)
+      result <- MigrateAction.liftIO(dataSource.inTransaction(action).io)
+    } yield result
+
+
+  final def getWorkspace(workspaceId: UUID): MigrateAction[Workspace] =
+    inTransaction {
+      _.workspaceQuery.findByIdOrFail(workspaceId.toString)
+    }
+
+
+  final def nowTimestamp: MigrateAction[Timestamp] =
+    MigrateAction.liftIO(IO {
+      Timestamp.valueOf(LocalDateTime.ofInstant(Instant.now, ZoneOffset.UTC))
+    })
+
+
+  final def randomSuffix(str: String): IO[String] = IO {
+    str ++ UUID.randomUUID.toString.replace("-", "")
+  }
+
+
+  final def noGoogleProjectError[A](migration: WorkspaceMigration): Throwable =
+    WorkspaceMigrationException(
+      message = "Workspace migration failed: Google Project not found.",
+      data = Map(
+        ("migrationId" -> migration.id),
+        ("workspaceId" -> migration.workspaceId),
+        ("googleProjectId" -> migration.newGoogleProjectId)
+      )
+    )
+
+
+  final def noWorkspaceBucketError[A](migration: WorkspaceMigration, bucket: GcsBucketName): Throwable =
+    WorkspaceMigrationException(
+      message = "Workspace migration failed: Workspace cloud bucket not found.",
+      data = Map(
+        ("migrationId" -> migration.id),
+        ("workspaceId" -> migration.workspaceId),
+        ("workspaceBucket" -> bucket)
+      )
+    )
+
+
+  final def noTmpBucketError[A](migration: WorkspaceMigration): Throwable =
+    WorkspaceMigrationException(
+      message = "Workspace migration failed: Temporary cloud storage bucket not found.",
+      data = Map(
+        ("migrationId" -> migration.id),
+        ("workspaceId" -> migration.workspaceId),
+        ("tmpBucket" -> migration.tmpBucketName)
+      )
+    )
```

Comment:
I wish we could also include the workspaceName in these 3 exceptions, but I think it would require a lot of rework or additional DB calls. But since this is in a unhappy path, maybe the performance doesn't really matter and the increased ease of debugging is worth it.

I think this is the unfortunate result of having a nonsense value as the workspaceId, which is annoying to search for in the db (removing the hyphens and then converting to hexadecimal) and can't be referenced via the UI.

Reply:
Good point - thank you!